compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -3057,6 +3057,8 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
         disjointnessBoundary(tp.effectiveBounds.hi)
       case tp: ErrorType =>
         defn.AnyType
+      case tp: NoType.type =>
+        defn.AnyType
```

Comment:
`defn.NothingType` is another alternative, but using `AnyType` gives error messages similar to what we get from Scala 3.3.